### ISSUE TITLE
Unable to Deploy to Tensorflow Serving

### ISSUE THREAD

**Issue by shba007** [ultralytics/ultralytics#2251]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

Response from tensorflow serving

{'error': 'Serving signature name: "serving_default" not found in signature def'}

### Environment

- YOLO v8
- Docker
- Tensorflow Serving 2.X

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hello, thank you for reaching out with your issue!

Based on the error message you provided, it seems that TensorFlow Serving is unable to find the "serving_default" signature in the SignatureDef of your exported model. This issue is usually caused by a mismatch between the signature name used during model export and the one expected by TensorFlow Serving.

To resolve this issue, I would recommend you to double-check the following steps:

Ensure that you have correctly exported your YOLOv8 model with the proper signature name ("serving_default") when using the TensorFlow export functionality.

Verify that the exported model contains the expected signature name. You can inspect the saved model by using the saved_model_cli tool that comes with TensorFlow. The command to run would be saved_model_cli show --dir /path/to/saved/model --all. This will display information about the model, including available signatures.

Make sure that you are using the correct path to the exported model when starting TensorFlow Serving. The path should point to the directory containing the saved_model.pb file and the variables directory.

**Comment by Codedrainer** [comment]
@shba007 have you solved this problem. i think i got caught with same problem. i have deployed model with docker tensorflow/serving but wont be able to infer because of "serving_default" is not present. i have used below command to export the model

yolo export model==best.pt format=saved_model imgsz=416 keras=True

@glenn-jocher would be great help if you can suggest some solutions

**Comment by glenn-jocher** [comment]
@Codedrainer it appears there might be a discrepancy in the export process. When exporting your YOLOv8 model to the TensorFlow SavedModel format, ensure you're following the correct procedure as outlined in our documentation. The `serving_default` signature should be automatically set during the export. If the issue persists, please verify the export settings and inspect the exported model to confirm the signature is included. If necessary, consult the `Export` section in our docs for detailed guidance.

**Comment by shba007** [comment]
Since it was a long time ago. I forgot how I exactly solved the problem

> NOTE: please change the paths according to you project structure

My guess is I have changed the model.conf file

model_config_list: {
  config: {
    name: "detector",
    base_path: "/saved_model/detector",
    model_platform: "tensorflow"
  }
}

If it does not work then you should also try to export the model using this code

`python

import tensorflow as tf
from tensorflow_hub import KerasLayer

model = tf.keras.models.load_model('../models/dev/object-detection/YOLO/(2023-05-06_14-40)_50')

input_shape = (640, 640, 3)

@tf.function(input_signature=[tf.TensorSpec(shape=(1, 640, 640, 3), dtype=tf.float32)])
def serving_fn(inputs):
    outputs = model(inputs)
    return outputs

preprocess_model = tf.keras.Sequential([
    tf.keras.layers.Rescaling(1./255, input_shape=input_shape),
    tf.keras.layers.Lambda(serving_fn)
])

preprocess_model.summary()

tf.keras.models.save_model(preprocess_model, '../models/deploy/servable/object-detection/detector', overwrite=True, include_optimizer=True, save_format="tf")
`

**Comment by Codedrainer** [comment]
@shba007 i did tried some solutions. i have loaded exported model (without signature) using `    loaded_model = tf.saved_model.load(input_model_path)
` and then save it with signature using `    tf.saved_model.save(loaded_model, output_model_path, signatures={'serving_default': serving_function.get_concrete_function()})
`

please find full conversion code below

`import tensorflow as tf

def load_and_save_model_with_signature(input_model_path, output_model_path):
    # Load the TensorFlow SavedModel
    loaded_model = tf.saved_model.load(input_model_path)

    # Define a serving function with the appropriate input signature
    @tf.function(input_signature=[tf.TensorSpec(shape=[1, 416, 416, 3], dtype=tf.float32, name='inputs_0')])
    def serving_function(inputs_0):
        # Process the input through the model
        outputs = loaded_model(inputs_0)
        return {'outputs': outputs}

    # Save the model with the new serving signature
    tf.saved_model.save(loaded_model, output_model_path, signatures={'serving_default': serving_function.get_concrete_function()})

    print(f"Model saved with serving signature at: {output_model_path}")

# Specify your model paths
input_model_path = 'yolov8/runs/detect/train2/weights/best_saved_model'
output_model_path = '/yolov8/runs/detect/train2/weights/saved_model'

# Run the function to load and save the model with the serving signature
load_and_save_model_with_signature(input_model_path, output_model_path)
`

after above operation i am able to deploy this on tensorflow/serving but getting output in tensors. Help me how can i achieve final result from output tensors as i am beginner in tensorflow/serving guide me. How can i get my classes and confidence score from output tensors

**Comment by Codedrainer** [comment]
@glenn-jocher  help me to solve above problem

**Comment by shba007** [comment]
@Codedrainer So I will try to explain it in steps
1. you send a api request at the endpoint of /detector:predict as a json { "instances": preprocessedImage }
2. you get a tensorflow of shape (1, 4+no of classes, x) where the 1 means you send one image x is the number of possible bounding boxes and 4 is x, y, width, height and rest is a one hot encoded array
3. now you have to filter by confidence you can use a boolean mask in tensorflow to do that
4. now do a NMS (Non-Max Suppression)
5. now optionally if you want to scale the boxes do it now
6. now format your responses according to your need and done

**Comment by Codedrainer** [comment]
@shba007 Thanks for brief explanantions, but i got my outputs like 

`dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 1
  }
  dim {
    size: 40
  }
  dim {
    size: 3549
  }
}
float_val: 4.0526714
float_val: 14.681505
float_val: 21.204597
float_val: 33.15705
float_val: 36.6662
float_val: 53.29041
float_val: 60.908257
float_val: 61.52846
float_val: 69.116516
float_val: 77.16216
float_val: 91.15697
float_val: 94.730865
float_val: 102.75214`

using below coed to infer my model from grpc

`import grpc
import tensorflow as tf
import numpy as np
from tensorflow_serving.apis import predict_pb2, prediction_service_pb2_grpc


def load_and_preprocess_image(image_path):
    # Load the image
    img = tf.io.read_file(image_path)
    img = tf.image.decode_image(img, channels=3)

    # Preprocess the image (adjust as per your model's requirement)
    img = tf.image.resize(img, [416, 416])  # Resize to model's input size
    img = img / 255.0  # Normalize if required by your model

    return img.numpy()


def make_prediction(image_path, server_url):
    # Load and preprocess the image
    image = load_and_preprocess_image(image_path)
    image_batch = np.expand_dims(image, axis=0)  # Reshape to (1, 416, 416, 3)

    # Create gRPC client and request
    channel = grpc.insecure_channel(server_url)
    stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)

    # Create PredictRequest
    request = predict_pb2.PredictRequest()
    request.model_spec.name = 'icai-captcha'
    request.model_spec.signature_name = 'serving_default'
    request.inputs['inputs_0'].CopyFrom(tf.make_tensor_proto(image_batch, shape=image_batch.shape))

    # Make the prediction
    response = stub.Predict(request, 10.0).outputs['outputs']  # 10 secs timeout
    return response


# Define the path to your image and the server URL
image_path = 'test/0a2dfc.jpg'
server_url = 'host_ip:10006'  # Example: 'localhost:

print(make_prediction(image_path, server_url))`

guide me if i am on right way to infer or not

**Comment by shba007** [comment]
The above steps I mentioned was for rest api, as I don't have much experience with grpc, I can't help you in this case. Although I can suggest you to do it in rest api first to do a sanity check then move on to grpc if you need

**Comment by Codedrainer** [comment]
@shba007 @glenn-jocher thanks for your suggestions. @shba007 can you help me to do with the rest api. Like how can I get information from those output tensors

**Comment by shba007** [comment]
A simple approch would be thinking the sending the preprocessed image and prediction from the tensorflow serving just as prediction  = model.predict(preprocessedImage)

you will do preprocessing and postprocessing same as the model is loaded locally once you have done that then just replace the prediction = model.predict(preprocessedImage) with a fetch request and capture the response as prediction now do all the post processing

**Comment by glenn-jocher** [comment]
@shba007 certainly! To handle the output tensors from the REST API, you'll follow these steps:

1. Send a preprocessed image to your TensorFlow Serving endpoint.
2. Receive the raw output tensor as a response.
3. Apply a threshold to filter out predictions with low confidence.
4. Perform non-maximum suppression (NMS) to eliminate overlapping boxes.
5. Apply softmax to the class probabilities to get confidence scores.
6. Extract the class with the highest score for each bounding box.

The post-processing steps are similar to what you would do if the model were loaded locally. Once you've set up the preprocessing and post-processing, you can replace the local model prediction call with an HTTP request to your TensorFlow Serving endpoint and process the response accordingly.

**Comment by Codedrainer** [comment]
@shba007 @glenn-jocher  I am stuck on this 2nd step you mentioned. Receive the raw output tensor as a response. Can you provide some documentation or article for this threshold and NMS implementation.

**Comment by shba007** [comment]
``` js
async function nms(boxes: [number, number, number, number, number, number][], maxOutputSize = 100, iouThreshold = 0.75, scoreThreshold = 0.5) {
	const boxesTensor = tf.tensor2d(boxes.map((box) => convertBox(box, DET_DIM, "CCWH", false, "XYXY", true)).map((box) => box.slice(0, 4)));
	const confsTensor = tf.tensor1d(boxes.map((box) => box[box.length - 2]));

	const selectedIndices = tf.image.nonMaxSuppression(
		boxesTensor,
		confsTensor,
		maxOutputSize,
		iouThreshold,
		scoreThreshold
	);
	const selectedBoxes = tf.gather(boxes, selectedIndices);
	return (await selectedBoxes.array()) as [number, number, number, number, number, number][];
}

/*
	format x_center, y_center, width, height in not normalized for 640, 640 image ->
	format x_center, y_center, width, height in normalized for self.dim image
 */
function scale(box: number[], dim: [number, number]) {
	let [xCenter, yCenter, width, height, ..._] = box;

	const factor = Math.max(...dim) / Math.max(...DET_DIM);
	const offset = Math.abs(dim[0] - dim[1]);

	xCenter *= factor;
	yCenter *= factor;
	width *= factor;
	height *= factor;

	if (dim[0] > dim[1])
		yCenter -= offset / 2;
	else if (dim[0] < dim[1])
		xCenter -= offset / 2;

	const scaledBox = [xCenter, yCenter, width, height];
	// console.log({ scaledBox });

	const convertedBox = convertBox([...scaledBox], [...dim], "CCWH", false, "CCWH", true);
	// console.log({ convertedBox });

	return convertedBox;
}


async function preprocess(image: Buffer): Promise<[number[], [number, number]]> {
	tf.engine().startScope()
	// Resize into 640x640
	const [resizedImageArray, dim] = await resize(image, DET_DIM)
	// console.log({ resizedImageArray })
	const reshapedImageArray = [[await reshape(resizedImageArray, DET_DIM)]] as unknown as number[]
	// console.log({ reshapedImageArray })
	tf.engine().startScope()

	return [reshapedImageArray, dim]
}

async function predict(image: Buffer): Promise<[number, number, number, number, number, number][]> {
	const config = useRuntimeConfig()

	const [preprocessedImage, dim] = await preprocess(image)

	try {
		const detections = await $fetch("/detector:predict", { baseURL: config.apiUrl, method: "POST", body: { "instances": preprocessedImage } })
		return postprocess(detections['predictions'][0], dim)
	} catch (error) {
		throw Error("Failed request Tensorflow Serving /detector:predict ", error)
	}
}

async function postprocess(predictions: [number, number, number, number, number, number][], dim: [number, number], confThreshold = 0.25): Promise<[number, number, number, number, number, number][]> {
	tf.engine().startScope()
	const [xs, ys, ws, hs, ...labels] = predictions

	const conf = tf.max(labels, 0)
	const labelIndex = tf.cast(tf.argMax(labels, 0), 'float32')
	const boxes = tf.stack([xs, ys, ws, hs, conf, labelIndex], 1)
	const mask = tf.greater(conf, confThreshold)

	// Filter by confidence Threshold
	const filteredBoxes = await (await tf.booleanMaskAsync(boxes, mask)).array() as [number, number, number, number, number, number][]
	// console.log({ filteredBoxes });

	// NMS
	const nmsBoxes = await nms(filteredBoxes)
	// console.log({ nmsBoxes });

	// Scale
	const scaledBoxes = nmsBoxes.map((nmsBox) => [...scale(nmsBox, dim), nmsBox[4], nmsBox[5]] as [number, number, number, number, number, number])
	// console.log({ scaledBoxes });
	tf.engine().startScope()

	return scaledBoxes
}

export default defineEventHandler<Detection>(async (event) => {
  const { image } = await readBody<{ image: string }>(event);
 const imageArray = await base64ToArray(image)
  const detections = await predict(imageArray)
  
  return {
	  "id": id,
	  "boxes": detections.map(([x, y, width, height, conf, label]) => {
		  if (label != 0)
			  return null
  
		  x *= 100
		  y *= 100
		  width *= 100
		  height *= 100
		  conf *= 100
  
		  return {
			  "x": x,
			  "y": y,
			  "width": width,
			  "height": height,
			  "conf": conf
		  }
	  }).filter((box) => box !== null)
  }
}
```
This is a code snippet for nitro which is a backend framework like express.js

**Comment by shba007** [comment]
Here is the repo from where I send the above snippet [AI](https://github.com/shba007/ai)

**Comment by glenn-jocher** [comment]
@shba007 thank you for sharing your repository. For those looking to implement thresholding and NMS, I recommend reviewing the documentation on our website, which provides guidance on these post-processing steps. While I can't review external code, the concepts you're exploring are relevant, and our documentation should help you apply these techniques effectively within the context of YOLOv8. 📚🚀
